openwisp_controller/subnet_division/handlers.py

```diff
@@ -0,0 +1,221 @@
+        modified_fields = {}
+        for field in db_instance._meta.fields:
+            instance_value = getattr(instance, field.name)
+            db_value = getattr(db_instance, field.name)
+            if instance_value != db_value:
+                modified_fields[field.name] = db_value
+
+        _subnet_division_rule_update_queue[str(instance.id)] = modified_fields
+
+
+@receiver(
+    post_save,
+    sender=SubnetDivisionRule,
+    dispatch_uid='update_subnet_division_rule_related',
+)
+def update_subnet_division_rule_related(sender, instance, created, **kwargs):
+    def create_subnet_division_index():
+        SubnetDivisionIndex.objects.create(
+            keyword=f'{instance.label}_prefixlen', rule_id=instance.id
+        )
+
+    def update_subnet_division_index_entries():
+        # Update related objects appropriately.
+        # NOTE: Currently only changing label and number of IPs is implemented/allowed.
+        try:
+            modified_fields = _subnet_division_rule_update_queue.pop(str(instance.id))
+        except KeyError:
+            return
+        else:
+            if 'label' in modified_fields:
+                tasks.update_subnet_division_index.delay(rule_id=str(instance.id))
+                tasks.update_subnet_name_description(rule_id=str(instance.id))
+            if 'number_of_ips' in modified_fields:
+                tasks.provision_extra_ips.delay(
+                    rule_id=str(instance.id),
+                    old_number_of_ips=modified_fields['number_of_ips'],
+                )
+
+    if created:
+        transaction.on_commit(create_subnet_division_index)
+    else:
+        transaction.on_commit(update_subnet_division_index_entries)
+
+
+@receiver(
+    post_delete, sender=SubnetDivisionRule, dispatch_uid='delete_provisioned_subnets'
+)
+def delete_provisioned_subnets(instance, **kwargs):
+    # Deleting an object of SubnetDivisionRule will automatically delete
+    # it's related SubnetDivisionIndex objects due to "on_delete=CASCADE".
+    # Similarly, deleting a Subnet object will automatically delete IpAddress
+    # objects related to it.
+    # Hence, this method only executes query of deleting Subnets provisioned by
+    # this corresponding SubnetDivisionRule.
+    def delete_subnets():
+        Subnet.objects.filter(
+            organization_id=instance.organization_id, name__startswith=instance.label
+        ).delete()
+
+    transaction.on_commit(delete_subnets)
+
+
+def provision_subnets_ips(instance, **kwargs):
+    def create_subnets_ips():
+        try:
+            organization_id = instance.config.device.organization_id
+            division_rule = instance.vpn.subnet.subnetdivisionrule_set.get(
+                organization_id__in=(organization_id, None),
+                type=(
+                    'openwisp_controller.subnet_division.types'
+                    '.VpnSubnetDivisionRuleType'
+                ),
+            )
```

Comment:
are you sure this query will always return only one result?

Can there be only one rule of a specific type and a specific org?

Part of this function is specific for the VPN rule, can you find a way to extract that part and move it to the vpn rule class?

For the rest of the function, can you find another place where to move this logic? For example a class method in one of the models could be good.

Reply:
> are you sure this query will always return only one result?

Yes, it should.

> Can there be only one rule of a specific type and a specific org?

Yes, I added following constraints as well. It also made logical sense. If someone wants that more subnets or IPs should be provisioned, they should update the rule instead of creating two rules of the same type.
https://github.com/openwisp/openwisp-controller/blob/8702de6d89f3423ed0c13c2dd260de928f2fb427/openwisp_controller/subnet_division/base/models.py#L36-L47